plugins/out_azure_kusto/azure_kusto.c

```diff
@@ -18,61 +18,266 @@
+ *
+ * Parameters:
+ *   - ctx: A pointer to the flb_azure_kusto context, which contains configuration and state
+ *          information necessary for the operation.
+ *   - client_id: A string representing the client ID for which the token is requested.
+ *
+ * Returns:
+ *   - A dynamically allocated string (flb_sds_t) containing the token if the request is successful.
+ *   - NULL if the request fails or if any error occurs during the process.
+ *
+ * The function performs the following steps:
+ *   1. Constructs the request URL using the Azure IMDS endpoint, API version, and resource.
+ *   2. Establishes a connection to the IMDS service using the upstream connection from the context.
+ *   3. Sends an HTTP GET request with the necessary headers, including Metadata and client_id.
+ *   4. Checks the HTTP response status and retrieves the token from the response payload if successful.
+ *   5. Cleans up resources, including the HTTP client, connection, and URL string, before returning.
+ *
+ * Note:
+ *   - The caller is responsible for freeing the returned token string.
+ *   - The function logs debug information about the HTTP request and response.
+ */
+flb_sds_t flb_azure_imds_get_token(struct flb_azure_kusto *ctx, flb_sds_t client_id)
 {
     int ret;
-    char *token;
-
-    /* Clear any previous oauth2 payload content */
-    flb_oauth2_payload_clear(ctx->o);
+    flb_sds_t url;
+    flb_sds_t response = NULL;
+    struct flb_http_client *client;
+    struct flb_connection *u_conn;
+    size_t b_sent;
 
-    ret = flb_oauth2_payload_append(ctx->o, "grant_type", 10, "client_credentials", 18);
-    if (ret == -1) {
-        flb_plg_error(ctx->ins, "error appending oauth2 params");
-        return -1;
+    url = flb_sds_create_size(256);
+    if (!url) {
+        flb_errno();
+        return NULL;
     }
 
-    ret = flb_oauth2_payload_append(ctx->o, "scope", 5, FLB_AZURE_KUSTO_SCOPE, 39);
-    if (ret == -1) {
-        flb_plg_error(ctx->ins, "error appending oauth2 params");
-        return -1;
-    }
+    flb_sds_printf(&url, "%s?api-version=%s&resource=%s",
+                   FLB_AZURE_IMDS_ENDPOINT,
+                   FLB_AZURE_IMDS_API_VERSION,
+                   FLB_AZURE_IMDS_RESOURCE);
 
-    ret = flb_oauth2_payload_append(ctx->o, "client_id", 9, ctx->client_id, -1);
-    if (ret == -1) {
-        flb_plg_error(ctx->ins, "error appending oauth2 params");
-        return -1;
+    u_conn = flb_upstream_conn_get(ctx->imds_upstream);
+    if (!u_conn) {
+        flb_sds_destroy(url);
+        return NULL;
     }
 
-    ret = flb_oauth2_payload_append(ctx->o, "client_secret", 13, ctx->client_secret, -1);
-    if (ret == -1) {
-        flb_plg_error(ctx->ins, "error appending oauth2 params");
-        return -1;
+    flb_plg_debug(ctx->ins, "[flb_azure_imds_get_token] client_id : %s", client_id);
+
+    client = flb_http_client(u_conn, FLB_HTTP_GET, url, NULL, 0, "169.254.169.254", 80, NULL, 0);
```

Comment:
validate that client is not NULL

Reply:
should we make this IP configurable and keep `169.254.169.254` as the default ? (same thing for the port)

Reply:
updated code